### DISCUSSION TITLE
Access Template's template_child with name

### DISCUSSION THREAD

**Discussion by eminfedar** [Relm4/Relm4#504] [Q&A] [answered]
I have a main template, which includes another template:
```rust
adw::Window {
    // ...

    #[template]
    MainWidget {

        #[template_child] // Can't access MainWidget->SettingsPage->btn_dark_mode like this
        btn_dark_mode {
            connect_clicked => Message::DarkMode,
        },

    },
}
```

Here is the MainWidget:
```rust
#[relm4::widget_template(pub)]
impl WidgetTemplate for MainWidget {
    view! {
        gtk::Box {
            set_orientation: gtk::Orientation::Vertical,

            gtk::Stack{
                    set_margin_all: 11,
                    set_margin_top: 7,

                    #[template]
                    add_child = &SettingsPage {} -> {
                        set_name: "settings",
                    },

                    #[template]
                    add_child = &MainPage {} -> {
                        set_name: "main",
                    },
                }

        }
    }
}
```

And inside of the `SettingsPage` template, there is a button which should send a Message:
```rust
#[relm4::widget_template(pub)]
impl WidgetTemplate for SettingsPage {
    view! {
        gtk::Box {
            set_orientation: gtk::Orientation::Vertical,
            set_spacing: 3,

            #[name = "btn_dark_mode"] // I NEED TO ACCESS THIS
            gtk::Button {

                #[wrap(Some)]
                set_child = &gtk::Image {
                    set_icon_name: Some("night-light-symbolic"), // When this clicked, I need to send Message::DarkMode
                },

            },
        }
    }
}
```

How can I do this?

**Comment by AaronErhardt** [comment, marked as answer]
I think this is currently not possible, at least in the macro.

Widget templates are relatively simple under the hood btw. For example, the generated struct for `SettingsPage` will roughly look like this:

```rust
struct SettingsPage {
    btn_dark_mode: gtk::Button,
}
```

All widgets and other templates without `#[name(...)]` are not accessible.

This means that this section in `MainWidget` makes `SettingsPage` inaccessible from outside because it has no name.

```rust
                    #[template]
                    add_child = &SettingsPage {} -> {
                        set_name: "settings",
                    },
```

To make is accessible, you need to use `#[name(...)]`, like this:

```rust
                    #[template]
                    #[name(settings_page)]
                    add_child = &SettingsPage {} -> {
                        set_name: "settings",
                    },
```

Then you can use `main_window.settings_page.btn_dark_mode` to access the widget. Unfortunately, the macro does not support nested template children (you can only reach `main_window.settings_page` but not further), but this should be easy to add. I will try to come up with a fix in the next few days.

**Reply by eminfedar** [reply]
Thanks a lot :) It would be a very nice feature to have this. Templates are really good at separating UI components to different files and views. Otherwise it is getting very big in a single view! macro.
